program/src/processor/distribute_rewards.rs

```diff
@@ -10,9 +17,75 @@ use crate::instruction::accounts::{Context, DistributeRewardsAccounts};
 /// 1. `[w]` Config account
 /// 2. `[]` System Program
 pub fn process_distribute_rewards(
-    _program_id: &Pubkey,
-    _ctx: Context<DistributeRewardsAccounts>,
-    _amount: u64,
+    program_id: &Pubkey,
+    ctx: Context<DistributeRewardsAccounts>,
+    amount: u64,
 ) -> ProgramResult {
+    // Account validation.
+
+    // config
+    // - owner must be the stake program
+    // - must be initialized
+
+    require!(
+        ctx.accounts.config.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "config"
+    );
+
+    let mut config_data = ctx.accounts.config.try_borrow_mut_data()?;
+    let config = bytemuck::try_from_bytes_mut::<Config>(&mut config_data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(
+        config.is_initialized(),
+        ProgramError::UninitializedAccount,
+        "config",
+    );
+
+    // payer
+    // - must be a signer
+
+    require!(
+        ctx.accounts.payer.is_signer,
+        ProgramError::MissingRequiredSignature,
+        "payer",
+    );
+
+    // system program
+    // - must be the system program
+
+    require!(
+        ctx.accounts.system_program.key == &system_program::ID,
+        ProgramError::IncorrectProgramId,
+        "system program"
+    );
+
+    // Calculates the rewards per token
+    //
+    // TODO: Do we still transfer the rewards to the config account if there are no
+    // staked tokens?
```

Comment:
I think you've done the right behavior of silently doing nothing if there are no staked tokens. It would be useful to log a message saying that the tokens were not transferred because nothing is staked.

Reply:
Unfortunately I don't think this is the right thing to do here.

If you have no staked tokens, then don't update the `stake_rewards_per_token` rate (since its 0 as a result of the 0 denominator), but still move the lamports. What if I want to feed rewards to the stake config while the staking ecosystem is still setting up, or not set up yet at all?

The rewards program doesn't prohibit the movement of reward lamports if the token supply is zero, it just doesn't update the rate.

https://github.com/paladin-bladesmith/rewards-program/blob/26845c684b485d7ea90c3332793afcd8316ef58b/program/src/processor.rs#L363-L385

Reply:
I was leaning towards that at first. If we do that, however, the pool is in a position where there are untouchable rewards. At which point, we, as the developers, need to change the program to make those rewards available. Rather than put ourselves in a position where we need to hack the program later, it's much easier for nothing to happen or for the program to throw an error, so that the rewards can be distributed later when something is staked.

Reply:
How does this affect the top-level system reward distribution then? If I need to send 10% to the governance treasury, 40% to stakers, and 50% to holders, and there are no staked tokens, then what happens to that 40%?

Reply:
Ah, that's an excellent point. Let's go with your suggestion then and transfer the tokens no matter what to be consistent

Reply:
Fwiw, the rewards only become untouchable when the system has no tokens, which would either be when it first launches or if everyone pulls their stake. In either case, I'd say the reward issuer should be aware of this and probably stop the funnel at the top.

Reply:
Same idea for distributing holder rewards with no tokens. Management concern.

Reply:
Ok, changed to transfer the rewards but not update the `accumulated_stake_rewards_per_token` rate.